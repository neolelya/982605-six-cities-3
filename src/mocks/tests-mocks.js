import {City, OFFER_FEATURES, RentalFeature, RentalType} from '../consts';

export const TEST_OFFERS = [
  {
    location: {
      city: City.AMSTERDAM,
      cityCoordinates: [52.38333, 4.9],
    },
    offers: [
      {
        id: 0,
        coordinates: [52.3909553943508, 4.85309666406198],
        rentalHost: {
          hostName: `Angelina`,
          hostAvatar: `img/avatar-angelina.jpg`,
          isSuper: true,
        },
        rentalTitle: `Beautiful & luxurious apartment at great location`,
        rentalImages: [
          `img/apartment-01.jpg`,
          `img/apartment-02.jpg`,
          `img/apartment-03.jpg`,
          `img/apartment-small-03.jpg`,
          `img/apartment-small-04.jpg`,
        ],
        rentalPrice: 120,
        rentalRating: 3.4,
        rentalType: RentalType.APARTMENT,
        isPremium: true,
        isBookmark: true,
        rentalDescription: [
          `A bright and charming apartment with 1 bedroom, located close to Museum Square in one of the most cosmopolitan and vibrant districts of Amsterdam "de PIJP".`,
          `Perfect for short holidays, business trips.`,
        ],
        rentalRoomsQuantity: 3,
        rentalMaxGuestsQuantity: 3,
        rentalFeatures: [RentalFeature.DRYER, RentalFeature.WASHINGMACHINE],
      },
      {
        id: 1,
        coordinates: [52.369553943508, 4.85309666406198],
        rentalHost: {
          hostName: `Max`,
          hostAvatar: `img/avatar-max.jpg`,
          isSuper: false,
        },
        rentalTitle: `Wood and stone place`,
        rentalImages: [
          `img/apartment-02.jpg`,
          `img/apartment-01.jpg`,
          `img/apartment-03.jpg`,
          `img/apartment-small-03.jpg`,
          `img/apartment-small-04.jpg`,
        ],
        rentalPrice: 80,
        rentalRating: 4.1,
        rentalType: RentalType.HOTEL,
        isPremium: false,
        isBookmark: true,
        rentalDescription: [
          `This package is special opportunity to connect accommodation in a traditional Dutch cottage located in a birch forest, taste the local food, and snowshoeing/hiking experience.`,
        ],
        rentalRoomsQuantity: 2,
        rentalMaxGuestsQuantity: 2,
        rentalFeatures: [
          RentalFeature.DRYER,
          RentalFeature.WASHINGMACHINE,
          RentalFeature.CABLETV,
          RentalFeature.KITCHEN,
          RentalFeature.KITCHEN,
        ],
      },
      {
        id: 2,
        coordinates: [52.3909553943508, 4.929309666406198],
        rentalHost: {
          hostName: `Angelina`,
          hostAvatar: `img/avatar-angelina.jpg`,
          isSuper: true,
        },
        rentalTitle: `Stylish and cozy place`,
        rentalImages: [
          `img/apartment-03.jpg`,
          `img/apartment-01.jpg`,
          `img/apartment-02.jpg`,
          `img/apartment-small-03.jpg`,
          `img/apartment-small-04.jpg`,
        ],
        rentalPrice: 132,
        rentalRating: 4.75,
        rentalType: RentalType.HOUSE,
        isPremium: true,
        isBookmark: false,
        rentalDescription: [
          `Spacious apartment situated in the city centre (oud-West) of Amsterdam.`,
          `Walking distance to the Jordaan, Museum Quarter, Anne Frank house and Leidseplein.`,
          `Supermarket right on the corner.`,
        ],
        rentalRoomsQuantity: 4,
        rentalMaxGuestsQuantity: 7,
        rentalFeatures: OFFER_FEATURES,
      },
      {
        id: 3,
        coordinates: [52.3809553943508, 4.939309666406198],
        rentalHost: {
          hostName: `Max`,
          hostAvatar: `img/avatar-max.jpg`,
          isSuper: false,
        },
        rentalTitle: `Excellent location and free parking`,
        rentalImages: [
          `img/room.jpg`,
          `img/room-small.jpg`,
          `img/apartment-01.jpg`,
          `img/apartment-02.jpg`,
          `img/apartment-03.jpg`,
          `img/apartment-small-03.jpg`,
          `img/apartment-small-04.jpg`,
        ],
        rentalPrice: 180,
        rentalRating: 4.5,
        rentalType: RentalType.APARTMENT,
        isPremium: false,
        isBookmark: false,
        rentalDescription: [
          `The studio apartments at Wittenberg have been carefully-designed to embrace and enhance the buildingâ€™s timeless authentic features, whilst incorporating striking yet traditional Dutch interiors.`,
        ],
        rentalRoomsQuantity: 2,
        rentalMaxGuestsQuantity: 4,
        rentalFeatures: [
          RentalFeature.DRYER,
          RentalFeature.WASHINGMACHINE,
          RentalFeature.PARKING,
        ],
      },
    ],
  },
];
