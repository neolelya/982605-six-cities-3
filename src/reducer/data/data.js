import {
  getOffersByCity,
  getUniqueCities,
  getUpdatedCurrentOffers,
  getUpdatedFavorites,
  getUpdatedOffers,
} from '../../utils';
import {
  AppRoute,
  City,
  OffersRestriction,
  ServerResponseStatusCode,
} from '../../consts';
import {offerAdapter, offersAdapter, reviewsAdapter} from '../../adapter';

const initialState = {
  allOffers: [],
  currentOffers: [],
  cities: [],
  reviews: [],
  nearbyOffers: [],
  isError: false,
  isSending: false,
  favorites: [],
};

const ActionType = {
  LOAD_OFFERS: `LOAD_OFFERS`,
  GET_OFFERS: `GET_OFFERS`,
  SET_ERROR: `SET_ERROR`,
  GET_REVIEWS: `GET_REVIEWS`,
  GET_NEARBY_OFFERS: `GET_NEARBY_OFFERS`,
  POST_REVIEW: `POST_REVIEW`,
  SET_SENDING: `SET_SENDING`,
  LOAD_FAVORITES: `LOAD_FAVORITES`,
  UPDATE_OFFER: `UPDATE_OFFER`,
};

const ActionCreator = {
  loadOffers: (offers) => ({
    type: ActionType.LOAD_OFFERS,
    payload: offers,
  }),
  getOffers: (city) => ({
    type: ActionType.GET_OFFERS,
    payload: city,
  }),
  setError: (isError) => ({
    type: ActionType.SET_ERROR,
    payload: isError,
  }),
  getReviews: (reviews) => ({
    type: ActionType.GET_REVIEWS,
    payload: reviews,
  }),
  getNearbyOffers: (nearbyOffers) => ({
    type: ActionType.GET_NEARBY_OFFERS,
    payload: nearbyOffers,
  }),
  postReview: (newReview) => ({
    type: ActionType.POST_REVIEW,
    payload: newReview,
  }),
  setSending: (isSending) => ({
    type: ActionType.SET_SENDING,
    payload: isSending,
  }),
  loadFavorites: (favorites) => ({
    type: ActionType.LOAD_FAVORITES,
    payload: favorites,
  }),
  updateOffer: (offer) => ({
    type: ActionType.UPDATE_OFFER,
    payload: offer,
  }),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_OFFERS:
      return Object.assign({}, state, {
        allOffers: action.payload,
        currentOffers: getOffersByCity(City.AMSTERDAM, action.payload),
        cities: getUniqueCities(action.payload),
      });

    case ActionType.GET_OFFERS:
      return Object.assign({}, state, {
        currentOffers: getOffersByCity(action.payload, state.allOffers),
      });

    case ActionType.SET_ERROR:
      return Object.assign({}, state, {
        isError: action.payload,
      });

    case ActionType.GET_REVIEWS:
      return Object.assign({}, state, {
        reviews: action.payload,
      });

    case ActionType.GET_NEARBY_OFFERS:
      return Object.assign({}, state, {
        nearbyOffers: action.payload,
      });

    case ActionType.POST_REVIEW:
      return Object.assign({}, state, {
        reviews: action.payload,
      });

    case ActionType.SET_SENDING:
      return Object.assign({}, state, {
        isSending: action.payload,
      });

    case ActionType.LOAD_FAVORITES:
      return Object.assign({}, state, {favorites: action.payload});

    case ActionType.UPDATE_OFFER:
      return Object.assign({}, state, {
        allOffers: getUpdatedOffers(action.payload, state.allOffers) || [],
        currentOffers: getUpdatedCurrentOffers(
            action.payload,
            state.currentOffers
        ),
        favorites: getUpdatedFavorites(action.payload, state.favorites),
      });
  }

  return state;
};

const Operation = {
  loadOffers: () => (dispatch, getState, api) => {
    return api
      .get(`/hotels`)
      .then((response) => {
        dispatch(
            ActionCreator.loadOffers(
                response.data.map((offer) => offersAdapter(offer))
            )
        );
      })
      .catch(() => {
        dispatch(ActionCreator.setError(true));
      });
  },

  getReviews: (id) => (dispatch, getState, api) => {
    return api
      .get(`/comments/${id}`)
      .then((response) => {
        dispatch(
            ActionCreator.getReviews(
                response.data.map((review) => reviewsAdapter(review))
            )
        );
      })
      .catch(() => {
        dispatch(ActionCreator.setError(true));
      });
  },

  getNearbyOffers: (id) => (dispatch, getState, api) => {
    return api.get(`/hotels/${id}/nearby`).then((response) => {
      dispatch(
          ActionCreator.getNearbyOffers(
              response.data
            .map((offer) => offersAdapter(offer))
            .slice(0, OffersRestriction.MAX_NEARBY_OFFERS_QUANTITY)
          )
      );
    });
  },

  postReview: (id, newReview) => (dispatch, getState, api) => {
    return api
      .post(`/comments/${id}`, newReview)
      .then((response) => {
        dispatch(ActionCreator.setSending(true));
        return response;
      })
      .then((response) => {
        dispatch(
            ActionCreator.getReviews(
                response.data.map((review) => reviewsAdapter(review))
            )
        );
      })
      .then(() => dispatch(ActionCreator.setSending(false)))
      .catch(() => {
        dispatch(ActionCreator.setError(true));
      });
  },

  loadFavorites: () => (dispatch, getState, api) => {
    return api.get(`/favorite`).then((response) => {
      dispatch(
          ActionCreator.loadFavorites(
              response.data.map((favorite) => offersAdapter(favorite))
          )
      );
    });
  },

  changeFavoriteStatus: (id, status) => (dispatch, getState, api) => {
    return api
      .post(`/favorite/${id}/${status}`)
      .then((response) => {
        dispatch(ActionCreator.updateOffer(offerAdapter(response.data)));
      })
      .catch((error) => {
        if (
          error.response &&
          error.response.status === ServerResponseStatusCode.UNAUTHORIZED
        ) {
          dispatch(ActionCreator.setError(true));
          window.location.pathname = AppRoute.LOGIN;
        }
      });
  },
};

export {reducer, ActionType, ActionCreator, Operation};
