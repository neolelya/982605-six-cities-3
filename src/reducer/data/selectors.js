import {NameSpace} from '../name-space';

export const getAllOffers = (state) => {
  return state[NameSpace.DATA].allOffers;
};

export const getCurrentOffers = (state) => {
  return state[NameSpace.DATA].currentOffers;
};

export const getCities = (state) => {
  return state[NameSpace.DATA].cities;
};

export const getIsError = (state) => {
  return state[NameSpace.DATA].isError;
};

export const getReviews = (state) => {
  return state[NameSpace.DATA].reviews;
};

export const getNearbyOffers = (state) => {
  return state[NameSpace.DATA].nearbyOffers;
};

export const getIsSending = (state) => {
  return state[NameSpace.DATA].isSending;
};

export const getFavorites = (state) => {
  const offers = state[NameSpace.DATA].favorites;

  const citiesMap = {};

  offers.forEach((offer) => {
    if (!citiesMap[offer.coordinates.city]) {
      citiesMap[offer.coordinates.city] = [];
    }
    citiesMap[offer.coordinates.city].push(offer);
    return citiesMap;
  });

  return Object.keys(citiesMap).map((city) => ({city, offers: citiesMap[city]}));
};

