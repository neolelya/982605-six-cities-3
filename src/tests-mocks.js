import {City, OFFER_FEATURES, RentalFeature, RentalType} from './consts';

export const TEST_OFFERS = [
  {
    location: {
      city: City.AMSTERDAM,
      cityCoordinates: [52.38333, 4.9],
    },
    offers: [
      {
        id: 0,
        coordinates: [52.3909553943508, 4.85309666406198],
        rentalHost: {
          hostName: `Angelina`,
          hostAvatar: `img/avatar-angelina.jpg`,
          isSuper: true,
        },
        rentalTitle: `Beautiful & luxurious apartment at great location`,
        rentalImages: [
          `img/apartment-01.jpg`,
          `img/apartment-02.jpg`,
          `img/apartment-03.jpg`,
          `img/apartment-small-03.jpg`,
          `img/apartment-small-04.jpg`,
        ],
        rentalPrice: 120,
        rentalRating: 3.4,
        rentalType: RentalType.APARTMENT,
        isPremium: true,
        isBookmark: true,
        rentalDescription: [
          `A bright and charming apartment with 1 bedroom, located close to Museum Square in one of the most cosmopolitan and vibrant districts of Amsterdam "de PIJP".`,
          `Perfect for short holidays, business trips.`,
        ],
        rentalRoomsQuantity: 3,
        rentalMaxGuestsQuantity: 3,
        rentalFeatures: [RentalFeature.DRYER, RentalFeature.WASHINGMACHINE],
        reviews: [
          {
            id: 0,
            name: `Adam`,
            avatar: `img/avatar-max.jpg`,
            rating: 4,
            date: new Date(2020, 0, 12),
            text: `Easy Check In: Management did extra mile, and allowed me to check in several hours before. Cleaning of my unit was swiftly organized, and properly introduced.`,
          },
          {
            id: 2,
            name: `Alex`,
            avatar: `img/avatar.svg`,
            rating: 5,
            date: new Date(2020, 1, 28),
            text: `We need a quick place to stay for the night and rented the same day. Was nice and clean. Had parking available for 15€ for the night. Was a little noisy but it ended by 11pm and we were able to get a decent night sleep. Had everything we needed.`,
          },
        ],
      },
      {
        id: 1,
        coordinates: [52.369553943508, 4.85309666406198],
        rentalHost: {
          hostName: `Max`,
          hostAvatar: `img/avatar-max.jpg`,
          isSuper: false,
        },
        rentalTitle: `Wood and stone place`,
        rentalImages: [
          `img/apartment-02.jpg`,
          `img/apartment-01.jpg`,
          `img/apartment-03.jpg`,
          `img/apartment-small-03.jpg`,
          `img/apartment-small-04.jpg`,
        ],
        rentalPrice: 80,
        rentalRating: 4.1,
        rentalType: RentalType.HOTEL,
        isPremium: false,
        isBookmark: true,
        rentalDescription: [
          `This package is special opportunity to connect accommodation in a traditional Dutch cottage located in a birch forest, taste the local food, and snowshoeing/hiking experience.`,
        ],
        rentalRoomsQuantity: 2,
        rentalMaxGuestsQuantity: 2,
        rentalFeatures: [
          RentalFeature.DRYER,
          RentalFeature.WASHINGMACHINE,
          RentalFeature.CABLETV,
          RentalFeature.KITCHEN,
          RentalFeature.KITCHEN,
        ],
        reviews: [
          {
            id: 0,
            name: `Adam`,
            avatar: `img/avatar-max.jpg`,
            rating: 4,
            date: new Date(2020, 0, 12),
            text: `Easy Check In: Management did extra mile, and allowed me to check in several hours before. Cleaning of my unit was swiftly organized, and properly introduced.`,
          },
          {
            id: 1,
            name: `Mary`,
            avatar: `img/avatar-angelina.jpg`,
            rating: 5,
            date: new Date(2020, 2, 15),
            text: `The building is in between the beach and metro station, 5 mins walking distance to both; 5 stops away from the Gothic Quarter by metro.`,
          },
          {
            id: 2,
            name: `Alex`,
            avatar: `img/avatar.svg`,
            rating: 5,
            date: new Date(2020, 1, 28),
            text: `We need a quick place to stay for the night and rented the same day. Was nice and clean. Had parking available for 15€ for the night. Was a little noisy but it ended by 11pm and we were able to get a decent night sleep. Had everything we needed.`,
          },
        ],
      },
      {
        id: 2,
        coordinates: [52.3909553943508, 4.929309666406198],
        rentalHost: {
          hostName: `Angelina`,
          hostAvatar: `img/avatar-angelina.jpg`,
          isSuper: true,
        },
        rentalTitle: `Stylish and cozy place`,
        rentalImages: [
          `img/apartment-03.jpg`,
          `img/apartment-01.jpg`,
          `img/apartment-02.jpg`,
          `img/apartment-small-03.jpg`,
          `img/apartment-small-04.jpg`,
        ],
        rentalPrice: 132,
        rentalRating: 4.75,
        rentalType: RentalType.HOUSE,
        isPremium: true,
        isBookmark: false,
        rentalDescription: [
          `Spacious apartment situated in the city centre (oud-West) of Amsterdam.`,
          `Walking distance to the Jordaan, Museum Quarter, Anne Frank house and Leidseplein.`,
          `Supermarket right on the corner.`,
        ],
        rentalRoomsQuantity: 4,
        rentalMaxGuestsQuantity: 7,
        rentalFeatures: OFFER_FEATURES,
        reviews: [
          {
            id: 2,
            name: `Alex`,
            avatar: `img/avatar.svg`,
            rating: 5,
            date: new Date(2020, 1, 28),
            text: `We need a quick place to stay for the night and rented the same day. Was nice and clean. Had parking available for 15€ for the night. Was a little noisy but it ended by 11pm and we were able to get a decent night sleep. Had everything we needed.`,
          },
        ],
      },
      {
        id: 3,
        coordinates: [52.3809553943508, 4.939309666406198],
        rentalHost: {
          hostName: `Max`,
          hostAvatar: `img/avatar-max.jpg`,
          isSuper: false,
        },
        rentalTitle: `Excellent location and free parking`,
        rentalImages: [
          `img/room.jpg`,
          `img/room-small.jpg`,
          `img/apartment-01.jpg`,
          `img/apartment-02.jpg`,
          `img/apartment-03.jpg`,
          `img/apartment-small-03.jpg`,
          `img/apartment-small-04.jpg`,
        ],
        rentalPrice: 180,
        rentalRating: 4.5,
        rentalType: RentalType.APARTMENT,
        isPremium: false,
        isBookmark: false,
        rentalDescription: [
          `The studio apartments at Wittenberg have been carefully-designed to embrace and enhance the building’s timeless authentic features, whilst incorporating striking yet traditional Dutch interiors.`,
        ],
        rentalRoomsQuantity: 2,
        rentalMaxGuestsQuantity: 4,
        rentalFeatures: [
          RentalFeature.DRYER,
          RentalFeature.WASHINGMACHINE,
          RentalFeature.PARKING,
        ],
        reviews: [],
      },
    ],
  },
];

export const TEST_REVIEWS = [
  {
    id: 0,
    name: `Adam`,
    avatar: `img/avatar-max.jpg`,
    rating: 4,
    date: new Date(2020, 0, 12),
    text: `Easy Check In: Management did extra mile, and allowed me to check in several hours before. Cleaning of my unit was swiftly organized, and properly introduced.`,
  },
  {
    id: 1,
    name: `Mary`,
    avatar: `img/avatar-angelina.jpg`,
    rating: 5,
    date: new Date(2020, 2, 15),
    text: `The building is in between the beach and metro station, 5 mins walking distance to both; 5 stops away from the Gothic Quarter by metro.`,
  },
  {
    id: 2,
    name: `Alex`,
    avatar: `img/avatar.svg`,
    rating: 5,
    date: new Date(2020, 1, 28),
    text: `We need a quick place to stay for the night and rented the same day. Was nice and clean. Had parking available for 15€ for the night. Was a little noisy but it ended by 11pm and we were able to get a decent night sleep. Had everything we needed.`,
  },
];

export const TEST_OFFERS_COORDINATES = [
  [52.3809553943508, 4.939309666406198],
  [52.3909553943508, 4.929309666406198],
  [52.3909553943508, 4.85309666406198],
  [52.369553943508, 4.85309666406198],
];
