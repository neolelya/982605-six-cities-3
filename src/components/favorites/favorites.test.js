import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {Provider} from 'react-redux';
import {createAPI} from '../../api';
import {MemoryRouter} from 'react-router-dom';
import Favorites from './favorites.jsx';
import {OFFERS} from '../../tests-mocks';
import {ActionType} from '../../reducer/data/data';

const api = createAPI();
const mockStore = configureStore([thunk.withExtraArgument(api)]);

const initialState = {
  DATA: {
    favorites: [],
  },
};

const expectedActions = [{type: ActionType.LOAD_FAVORITES}];

const store = mockStore(initialState, expectedActions);

it(`Should Favorites render correctly with favoritesList data`, () => {
  const tree = renderer
    .create(
        <MemoryRouter>
          <Provider store={store}>
            <Favorites
              userEmail={`someEmail@mail.su`}
              favorites={OFFERS[0]}
              onRentalCardHover={() => {}}
              onBookmarkClick={() => {}}
              onFavoritesLoad={() => {}}
            />
          </Provider>
        </MemoryRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
