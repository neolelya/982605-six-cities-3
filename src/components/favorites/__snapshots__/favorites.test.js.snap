// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Favorites render correctly with favoritesList data 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link header__logo-link--active"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/favorites"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  className="header__user-name user__name"
                  onClick={[Function]}
                >
                  someEmail@mail.su
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--favorites page__main--favorites-empty"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites favorites--empty"
      >
        <h1
          className="visually-hidden"
        >
          Favorites (empty)
        </h1>
        <div
          className="favorites__status-wrapper"
        >
          <b
            className="favorites__status"
          >
            Nothing yet saved.
          </b>
          <p
            className="favorites__status-description"
          >
            Save properties to narrow down search or plan yor future trips.
          </p>
        </div>
      </section>
    </div>
  </main>
  <footer
    className="footer"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;
